lib/forms.js

```diff
@@ -11,55 +11,136 @@ var deleteAppReferences = require('./impl/deleteAppRefrences.js');
 var dataSourceImpl = require('./impl/dataSources');
 var dataTargetImpl = require('./impl/dataTargets');
 var config = require('./config');
+var mongoUriParser = require('mongodb-uri');
 var logger = logfns.getLogger();
+var async = require('async');
+
+//Single mongoose connection for forms.
+var mongooseConnection;
+var dbConnection;
+
+//TODO: This should be global.
+mongoose.Promise = global.Promise;
+
+
+function getMongooseConnection(uri, connections, cb) {
+  //Already have a mongoose connection, going to re-use the connection pool to switch to the required database.
+  var parsedMongoUri = mongoUriParser.parse(uri);
+
+  parsedMongoUri.options = parsedMongoUri.options || {};
+
+  //TODO: This can be configured.
+  parsedMongoUri.options.poolSize = 5;
+
+  var newUrl = mongoUriParser.format(parsedMongoUri);
+
+  //If there is no mongoose connection, then create one at least.
+  if (!mongooseConnection) {
+    logger.debug("No Mongoose Connection, creating one.");
+    mongooseConnection = mongoose.createConnection(newUrl);
+  }
+
+  //There is already a mongoose connection, can already use this.
+  if (connections.mongooseConnection) {
+    logger.debug("Using cached mongoose Db connection");
+    return cb(undefined, connections.mongooseConnection);
+  }
+
+  var dbName = parsedMongoUri.database;
+
+  //Already have a mongoose connection, need to get a handle to the other database.
+  var mongooseDbConnection = mongooseConnection.useDb(dbName);
+
+  logger.debug("Mongoose connected to new db");
+  //Authenticating to the database
+
+  connections.mongooseConnection = mongooseDbConnection;
+
+  //Db is authenticated, can now do stuff.
+  return cb(undefined, mongooseDbConnection);
+}
 
 
 var forms = {
   connections: {},
   setConfig: config.set,
 
-  initConnection : function(options, cb) {
+  initConnection : function(options, callback) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Function has a complexity of 21 which is greater than 20 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)